### ISSUE TITLE
Deserializing nested objects (here: SeedGenerator as seed for GlorotUniform initializer)

### ISSUE THREAD

**Issue by bruehle** [keras-team/keras#20209]
Hi, I am trying to load a custom model that uses a GlorotUniform kernel initializer that was seeded with a seed generator. When trying to load the model, an error occurs that is due to the seed argument of the GlorotUniform initializer not deserializing properly (it is still a dict instead of a SeedGenerator instance).

Minimal example to reproduce (Keras 3.5.0, Python 3.11.6):
```
import keras

kernel_init = keras.initializers.GlorotUniform(keras.random.SeedGenerator(0))
print(type(kernel_init.seed))  # <class 'keras.src.random.seed_generator.SeedGenerator'>
kernel_init_serialized = keras.saving.serialize_keras_object(kernel_init)
kernel_init_deserialized = keras.saving.deserialize_keras_object(kernel_init_serialized)
print(type(kernel_init_deserialized.seed))  # <class 'dict'>
```

Currently I am deserializing the seed generator manually as a workaround, like so:
```
kernel_init_deserialized.seed = keras.saving.deserialize_keras_object(kernel_init_deserialized.seed)
print(type(kernel_init_deserialized.seed))  # <class 'keras.src.random.seed_generator.SeedGenerator'>
```

It works (I think), but I still would have expected such nested constructs to be deserialized automatically without the workaround.

**Comment by ghsanti** [comment]
I'm not sure why, but I think [that's not supported](https://github.com/keras-team/keras/issues/19796#issuecomment-2148209903).

> The Keras saving format isn't currently set up to save and load random seeds. That's not part of the API contract. It's something that could be added, though.
> For now you can simply make your Dropout layer deterministic by passing seed=... (an int) to the constructor.

**Comment by bruehle** [comment]
@ghsanti 
Thank you very much for your response. I think you are right, that is probably the problem. The reason I was a bit confused about this is that in the keras documentation on [Save, serialize, and export models](https://keras.io/guides/serialization_and_saving/) in the section Appendix: Handling custom objects/Defining the config methods, there is actually an example with a kernel initializer, and it says:

> Note that overriding from_config is unnecessary above for MyDense because hidden_units, kernel_initializer, and kernel_regularizer are ints, strings, and **a built-in Keras object**, respectively. This means that the default from_config implementation of cls(**config) will work as intended.

So I just assumed that all built-in Keras objects will be handled automatically, including the seed generators. And since it serializes correctly and without a warning, and can even be manually deserialized again, I found this behavior odd.

Unfortunately, using an `int` as a seed has the disadvantage that all calls to the initializer will produce the exact same values, which I wanted to avoid:

```python
import keras
import numpy as np

def build_submodel(kernel_init):
    inp = keras.layers.Input(shape=(256, 256, 1))
    x = keras.layers.Conv2D(32, (3, 3), kernel_initializer=kernel_init)(inp)
    model = keras.models.Model(inputs=inp, outputs=x)
    return model

seed = 0
submodel_1 = build_submodel(keras.initializers.GlorotUniform(seed))
submodel_2 = build_submodel(keras.initializers.GlorotUniform(seed))
print(np.all([np.all(submodel_1.trainable_weights[i].numpy() == submodel_2.trainable_weights[i].numpy()) for i in range(len(submodel_1.trainable_weights))]))
# True

seed = keras.random.SeedGenerator(0)
submodel_1 = build_submodel(keras.initializers.GlorotUniform(seed))
submodel_2 = build_submodel(keras.initializers.GlorotUniform(seed))
print(np.all([np.all(submodel_1.trainable_weights[i].numpy() == submodel_2.trainable_weights[i].numpy()) for i in range(len(submodel_1.trainable_weights))]))
# False
```

**Comment by ghsanti** [comment]
It may be possible to just deserialise using `from_config()` and the method you used first, or some variation.

Another solution seems this one?

This produces same seed per program. However, the seed isn't stored, and am unsure it's platform independent.

```py
import keras
import numpy as np

keras.utils.set_random_seed(812)

def build_submodel():
    inp = keras.layers.Input(shape=(256, 256, 1))
    x = keras.layers.Conv2D(32, (3, 3))(inp)
    model = keras.models.Model(inputs=inp, outputs=x)
    return model


submodel_1 = build_submodel()
submodel_2 = build_submodel()
m1=submodel_1.trainable_weights[0].numpy() 
m2=submodel_2.trainable_weights[0].numpy()
print(m1,m2)
```

**Comment by bruehle** [comment]
Thank you for your help. Yes, I currently use the method I mentioned first, and it seems to work. I can save and load the model correctly, and the kernels were initialized with different values before saving. It's just not very pretty:

```python
import keras
import numpy as np

def build_submodel(kernel_init):
    inp = keras.layers.Input(shape=(256, 256, 1))
    x = keras.layers.Conv2D(32, (3, 3), kernel_initializer=kernel_init)(inp)
    model = keras.models.Model(inputs=inp, outputs=x)
    return model

@keras.saving.register_keras_serializable()
class CustomModel(keras.Model):
    def __init__(self, submodel_1, submodel_2, **kwargs):
        super(CustomModel, self).__init__(**kwargs)
        self.submodel_1 = submodel_1
        self.submodel_2 = submodel_2
        self.built = True

    def get_config(self):
        config = super().get_config()
        config.update({'submodel_1': self.submodel_1, 'submodel_2': self.submodel_2})
        return config

    @classmethod
    def from_config(cls, config):
        # Manually deserialize seed generator
        for net in ('submodel_1', 'submodel_2'):
            for i in range(0, len(config[net]['config']['layers'])):
                if 'kernel_initializer' in config[net]['config']['layers'][i]['config'].keys():
                    kernel_init = keras.saving.deserialize_keras_object(config[net]['config']['layers'][i]['config']['kernel_initializer']['config']['seed'])
                    config[net]['config']['layers'][i]['config']['kernel_initializer']['config']['seed'] = kernel_init

        submodel_1 = keras.saving.deserialize_keras_object(config.pop('submodel_1'))
        submodel_2 = keras.saving.deserialize_keras_object(config.pop('submodel_2'))
        return cls(submodel_1, submodel_2, **config)

seed = 0
submodel_1 = build_submodel(keras.initializers.GlorotUniform(seed))
submodel_2 = build_submodel(keras.initializers.GlorotUniform(seed))
my_custom_model = CustomModel(submodel_1, submodel_2)

my_custom_model.save('my_model.keras')
model_reloaded = keras.models.load_model('my_model.keras')

w1 = model_reloaded.submodel_1.trainable_weights[0].numpy()
w2 = model_reloaded.submodel_2.trainable_weights[0].numpy()
print(np.all(w1 == w2))
# True

seed = keras.random.SeedGenerator(0)
submodel_1 = build_submodel(keras.initializers.GlorotUniform(seed))
submodel_2 = build_submodel(keras.initializers.GlorotUniform(seed))
my_custom_model = CustomModel(submodel_1, submodel_2)

my_custom_model.save('my_model.keras')
model_reloaded = keras.models.load_model('my_model.keras')

w1 = model_reloaded.submodel_1.trainable_weights[0].numpy()
w2 = model_reloaded.submodel_2.trainable_weights[0].numpy()
print(np.all(w1 == w2))
# False
```

But using `keras.utils.set_random_seed(812)` and an unseeded `keras.initializers.GlorotUniform()` might be a good alternative. Thank you for the suggestion.

**Comment by fchollet** [comment]
What you're trying to do is legit and should be supported. I just fixed the deserialization issue. The serialization roundtrip works at HEAD and will work in the next release.
